### PR TITLE
Fixed Issue #17295 : Search REST API returns wrong total_count

### PR DESCRIPTION
**Pull request by ronak2ram**
### Description (*)
I added function for getting total_count of result item.

### Fixed Issues (if relevant)
1. magento/magento2#17295: Search REST API returns wrong total_count

### Request :
 `rest/V1/search?searchCriteria[requestName]=quick_search_container&searchCriteria[filterGroups][0][filters][0][field]=search_term&searchCriteria[filterGroups][0][filters][0][conditionType]=like&searchCriteria[filterGroups][0][filters][0][value]=test&searchCriteria[current_page]=1&searchCriteria[page_size]=2`

### Response : 
```
{
    "items": [
        {
            "id": 126,
            "custom_attributes": [
                {
                    "attribute_code": "score",
                    "value": "73.7660923004150400"
                }
            ]
        },
        {
            "id": 61,
            "custom_attributes": [
                {
                    "attribute_code": "score",
                    "value": "73.7660923004150400"
                }
            ]
        }
    ],
    "aggregations": {
        "buckets": [
            {
                "name": "category_bucket",
                "values": [
                    {
                        "value": "32",
                        "metrics": [
                            "32",
                            "1"
                        ]
                    }
                ]
            },
            {
                "name": "stock_status_bucket",
                "values": []
            },
            {
                "name": "rating_summary_bucket",
                "values": []
            },
            {
                "name": "am_is_new_bucket",
                "values": []
            },
            {
                "name": "am_on_sale_bucket",
                "values": []
            },
            {
                "name": "color_bucket",
                "values": []
            },
            {
                "name": "sub_category_bucket",
                "values": []
            }
        ],
        "bucket_names": [
            "category_bucket",
            "stock_status_bucket",
            "rating_summary_bucket",
            "am_is_new_bucket",
            "am_on_sale_bucket",
            "color_bucket",
            "sub_category_bucket"
        ]
    },
    "search_criteria": {
        "request_name": "quick_search_container",
        "filter_groups": [
            {
                "filters": [
                    {
                        "field": "search_term",
                        "value": "test",
                        "condition_type": "like"
                    }
                ]
            }
        ],
        "page_size": 2,
        "current_page": 1
    },
    "total_count": 9
}
```
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @ronak2ram. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by ronak2ram** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @ronak2ram. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @ronak2ram, here is your [new Magento instance](https://pr-20253.instances.magento-community.engineering).
Admin access: [https://pr-20253.instances.magento-community.engineering/admin](https://pr-20253.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by ronak2ram** [comment]
Confirmed on the test instance.

**Comment by davidverholen** [comment]
@ronak2ram  currently your solution introduces a huge performance issue by sending a second search on each request to the search engine. 

I did not go into detail yet, but wouldn't it be possible, to resolve this issue by fixing the original call to the search engine so it calculates the correct total count?

File: lib/internal/Magento/Framework/Search/Search.php [hunk 1]
```diff
@@ -97,4 +97,40 @@ private function addFieldToFilter($field, $condition = null)
...
         return $this;
     }
+
+    /**
+     * @param SearchCriteriaInterface $searchCriteria
+     * @return int
+     */
+    public function getTotalCount(SearchCriteriaInterface $searchCriteria)
+    {
+        $requestBuilder = \Magento\Framework\App\ObjectManager::getInstance()
```
Comment (davidverholen) [inline lib/internal/Magento/Framework/Search/Search.php]: ObjectManager must not be used directly here

File: lib/internal/Magento/Framework/Search/Search.php [hunk 2]
```diff
@@ -97,4 +97,40 @@ private function addFieldToFilter($field, $condition = null)
...
+                    }
+                    if (!empty($condition['to'])) {
+                        $requestBuilder->bind("{$field}.to", $condition['to']);
+                    }
+                }
+            }
+        }
+
+        $request = $requestBuilder->create();
+        $searchResponse = $this->searchEngine->search($request);
```
Comment (davidverholen) [inline lib/internal/Magento/Framework/Search/Search.php]: the search engine should not require a second search just to figure out the correct count

**Review by davidverholen — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/Search/Search.php (hunk 1 above)
Comment (ronak2ram) [reply lib/internal/Magento/Framework/Search/Search.php]: @davidverholen I will remove ObjectManager in next commit.

File: lib/internal/Magento/Framework/Search/Search.php (hunk 2 above)
Comment (ronak2ram) [reply lib/internal/Magento/Framework/Search/Search.php]: I do not find any other solution can you please suggest me a better solution to eliminate searchEngine?

**Comment by ronak2ram** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @ronak2ram. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @ronak2ram, here is your [new Magento instance](https://pr-20253.instances.magento-community.engineering).
Admin access: [https://pr-20253.instances.magento-community.engineering/admin](https://pr-20253.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by ronak2ram** [comment]
Hello @davidverholen @orlangur 
Can you please review this pull. I changed the flow to the resolved issue.

**Comment by davidverholen** [comment]
Hi @ronak2ram , 

this solution seems to be more "On the point" since it mostly extends the mysql search adapter to reflect the correct total count (even if the page size is lower than the total count).

However, you are introducing a new public method in the Query Response of the search framework which then probably needs to be implemented in all additional Search engines (like elasticsearch and 3rd party modules).

I will try to investigate a bit deeper if I see a solution, without changing the public api of the search framework. If not, it seems to be a design problem of the search framework api, which seems to be missing exactly the method you introduced in your change.

**Comment by ronak2ram** [comment]
Hello @davidverholen 
Is there any update regarding this PR?

File: lib/internal/Magento/Framework/Search/Adapter/Mysql/Adapter.php [hunk 3]
```diff
@@ -49,6 +49,16 @@ class Adapter implements AdapterInterface
      */
     private $temporaryStorageFactory;
 
+    /**
+     * Query Select Parts to be skipped when prepare query for count
+     *
+     * @var array
+     */
+    protected $countSqlSkipParts = [
```
Comment (sivaschenko) [inline lib/internal/Magento/Framework/Search/Adapter/Mysql/Adapter.php:57]: Please change the access modifier to private

File: lib/internal/Magento/Framework/Search/Response/QueryResponse.php [hunk 4]
```diff
@@ -65,4 +74,12 @@ public function getAggregations()
     {
         return $this->aggregations;
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTotalCount()
```
Comment (sivaschenko) [inline lib/internal/Magento/Framework/Search/Response/QueryResponse.php:81]: @buskamuza @melnikovi can you please take a look if we can allow adding this method to the framework. It appears there is no other way to return the total count of the search result.

File: lib/internal/Magento/Framework/Search/Adapter/Mysql/Adapter.php [hunk 5]
```diff
@@ -114,4 +125,35 @@ private function getConnection()
...
+        $totalRecords = $this->getConnection()->fetchOne($parentSelect);
+        return (int)$totalRecords;
+    }
+
+    /**
+     * Reset limit and offset
+     *
+     * @return Select
+     */
+    protected function getSelectCountSql($query)
```
Comment (sivaschenko) [inline lib/internal/Magento/Framework/Search/Adapter/Mysql/Adapter.php:150]: Please change method access modifier to private

File: lib/internal/Magento/Framework/Search/Adapter/Mysql/Adapter.php [hunk 6]
```diff
@@ -114,4 +125,35 @@ private function getConnection()
     {
         return $this->resource->getConnection();
     }
+
+    /**
+     * Get rows size
+     *
+     * @return int
+     */
+    public function getSize($query)
```
Comment (sivaschenko) [inline lib/internal/Magento/Framework/Search/Adapter/Mysql/Adapter.php:134]: Please change method access modifier to private

File: lib/internal/Magento/Framework/Search/Response/QueryResponse.php [hunk 7]
```diff
@@ -29,14 +29,23 @@ class QueryResponse implements ResponseInterface
      */
     protected $aggregations;
 
+    /**
+     * Total count of collection
+     *
+     * @var int
+     */
+    protected $totalCount;
```
Comment (sivaschenko) [inline lib/internal/Magento/Framework/Search/Response/QueryResponse.php:37]: Please change property access modifier to private

File: lib/internal/Magento/Framework/Search/Response/QueryResponse.php [hunk 8]
```diff
@@ -29,14 +29,23 @@ class QueryResponse implements ResponseInterface
...
+
     /**
      * @param Document[] $documents
      * @param AggregationInterface $aggregations
+     * @param Total size int $size
```
Comment (sivaschenko) [inline lib/internal/Magento/Framework/Search/Response/QueryResponse.php:42]: Please correct the phpdoc: `@param int $size Total size`

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @ronak2ram thanks for the pull request, please see my code review comments.

@buskamuza @melnikovi please review this contribution, in order to process it we'll need an approval of adding \Magento\Framework\Search\Response\QueryResponse::getTotalCount public method.

File: lib/internal/Magento/Framework/Search/Response/QueryResponse.php (hunk 4 above)
Comment (melnikovi) [reply lib/internal/Magento/Framework/Search/Response/QueryResponse.php:81]: This class has count method already, that seem to have similar purpose. Are there cases when count that it uses is not efficient?

Comment (ronak2ram) [reply lib/internal/Magento/Framework/Search/Response/QueryResponse.php:81]: @melnikovi Count method returns a total count of the search result items of the current page.
We make getTotalCount for the return all pages items count.

**Comment by ronak2ram** [comment]
Hello Masters @sivaschenko @melnikovi @davidverholen 
Thanks for the review of this PR.
By mistake, I deleted the base branch of this pull request. So I can't able to update the changes in this PR.
So I created new PR with the suggested changes by @sivaschenko and latest code.

Please Review PR : #21713

**Comment by sivaschenko** [comment]
Thanks @ronak2ram ! As all the concerns in the review were fixed, I approved the new pull request, closing this one as outdated.

**Pull request closed without merging**